### PR TITLE
Feature: Validate Input

### PR DESCRIPTION
**Pull request by maaslalani**
Based on https://github.com/charmbracelet/bubbles/pull/114

Adds `Validate` property to `textinput` to ensure input is valid.

Validate is a function that accepts a `string` and returns an `error` which may be `nil` if the string is valid.

```go
// ValidateFunc is a function that returns an error if the input is invalid.
type ValidateFunc func(string) error
```

### CODE REVIEW

File: textinput/textinput.go [hunk 1]
```diff
@@ -180,17 +188,30 @@ func New() Model {
...
 // SetValue sets the value of the text input.
-func (m *Model) SetValue(s string) {
+// Returns an error if validation fails.
+func (m *Model) SetValue(s string) error {
+	if m.Validate != nil {
+		err := m.Validate(s)
+		if err != nil {
+			m.Err = err
+			return err
+		}
```
Comment (caarlos0) [inline textinput/textinput.go:198]: nickpicking, I would do:

```suggestion
		if err := m.Validate(s); err != nil {
			m.Err = err
			return err
		}
```

File: textinput/textinput.go [hunk 2]
```diff
@@ -180,17 +188,30 @@ func New() Model {
 var NewModel = New
 
 // SetValue sets the value of the text input.
-func (m *Model) SetValue(s string) {
+// Returns an error if validation fails.
+func (m *Model) SetValue(s string) error {
```
[the remark this answers is not in the document]
Comment (maaslalani) [reply textinput/textinput.go]: You're absolutely right. Fixed :)

**Review by caarlos0 — APPROVED** [review]

File: textinput/textinput.go [hunk 3]
```diff
@@ -181,13 +190,22 @@ var NewModel = New
...
+	m.Err = nil
+
 	runes := []rune(s)
 	if m.CharLimit > 0 && len(runes) > m.CharLimit {
 		m.value = runes[:m.CharLimit]
 	} else {
 		m.value = runes
 	}
-	if m.pos == 0 || m.pos > len(m.value) {
+	if (m.pos == 0 && len(m.value) == 0) || m.pos > len(m.value) {
```
[the remark this answers is not in the document]
Comment (maaslalani) [reply textinput/textinput.go:208]: Hey! That's definitely a bug, I will fix it soon! Thank you for catching it.

[the remark this answers is not in the document]
Comment (maaslalani) [reply textinput/textinput.go:208]: If you would like to, go for it! It would be much appreciated :)

**Pull request merged**